// bionode-go/lib/go_gene_annotation.js
var through    = require('through2');
var mysql      = require('mysql');
var connection = undefined;
var maxBatchTimeout = 1000;
var maxBatchSize = 20;

var GO = exports

// call this function to initiate a database connection with the public server at EBI
// set different batch query options where needed.
GO.start_session = function(fastOptions) {
	connection = mysql.createConnection({
	    host     : 'mysql.ebi.ac.uk',
	    user     : 'go_select',
	    database : 'go_latest',
	    password : 'amigo',
	    port     : 4085
	  });
	connection.connect();
}

// call this function for good practice to formally end the database connection
GO.end_session = function () {
	connection.end();
}

// call this function with a geneID and source DB for getting all direct gene annotations
// http://amigo.geneontology.org/goose provides some more help until I get around to full doc
GO.direct_annotation = through.obj(function transform(obj, enc, next) {
  var self = this;
	var gene = obj.gene
	var genus = obj.genus
	var query = "SELECT term.name AS superterm_name, term.acc AS superterm_acc, term.term_type AS superterm_type, association.*, gene_product.symbol AS gp_symbol, gene_product.symbol AS gp_full_name, dbxref.xref_dbname AS gp_dbname, dbxref.xref_key AS gp_acc, species.genus, species.species, species.ncbi_taxa_id, species.common_name FROM term INNER JOIN graph_path ON (term.id=graph_path.term1_id) INNER JOIN association ON (graph_path.term2_id=association.term_id) INNER JOIN gene_product ON (association.gene_product_id=gene_product.id) INNER JOIN species ON (gene_product.species_id=species.id) INNER JOIN dbxref ON (gene_product.dbxref_id=dbxref.id) WHERE gene_product.symbol = '"+gene+"' AND species.genus = '"+genus+"';"

	connection.query(query, function(err, rows, fields) {
    if (err) return self.emit('error', err);
    rows.forEach(function(row) {
      self.push(row);
      //console.log('The solution is: ', rows[0]);
    });
  });
  next();
})
